helios-tools/src/main/java/com/spotify/helios/cli/command/HostListCommand.java

```diff
@@ -62,206 +51,233 @@
-          } else {
-            memUsage = cpus = mem = loadAvg = os = docker = "";
-          }
-
-          final String version;
-          if (s.getAgentInfo() != null) {
-            version = Optional.fromNullable(s.getAgentInfo().getVersion()).or("");
-          } else {
-            version = "";
-          }
-
-          String status = s.getStatus() == UP ? "Up" : "Down";
-          if (s.getAgentInfo() != null) {
-            final long startTime = s.getAgentInfo().getStartTime();
-            final long upTime = s.getAgentInfo().getUptime();
-            if (s.getStatus() == UP) {
-              status += " " + humanDuration(currentTimeMillis() - startTime);
-            } else {
-              status += " " + humanDuration(currentTimeMillis() - startTime - upTime);
-            }
-          }
-
-          final String hostLabels = Joiner.on(", ").withKeyValueSeparator("=").join(s.getLabels());
-
-          table.row(formatHostname(full, host), status, s.getJobs().size(),
-              runningDeployedJobs.size(), cpus, mem, loadAvg, memUsage, os, version, docker,
-              hostLabels);
-        }
-
-        table.print();
-      }
-    }
-    return 0;
-  }
+	private final Argument quietArg;
+	private final Argument patternArg;
+	private final Argument fullArg;
+	private final Argument statusArg;
+	private final Argument labelsArg;
+
+	private final String statusChoicesString;
+
+	public HostListCommand(final Subparser parser) {
+		super(parser);
+
+		Collection<String> statusChoices = Collections2.transform(Arrays.asList(HostStatus.Status.values()),
+				new Function<HostStatus.Status, String>() {
+					@Override
+					public String apply(final HostStatus.Status input) {
+						return input.toString();
+					}
+				});
+
+		statusChoicesString = Joiner.on(", ").join(statusChoices);
+
+		parser.help("list hosts");
+
+		patternArg = parser.addArgument("pattern").nargs("?").setDefault("").help("Pattern to filter hosts with");
+
+		quietArg = parser.addArgument("-q").action(storeTrue()).help("only print host names");
+
+		fullArg = parser.addArgument("-f").action(storeTrue()).help("Print full host names.");
+
+		statusArg = parser.addArgument("--status").nargs("?")
+				.choices(statusChoices.toArray(new String[statusChoices.size()]))
+				.help("Filter hosts by its status. Valid statuses are: " + statusChoicesString);
+
+		labelsArg = parser.addArgument("-l", "--labels").action(append()).setDefault(new ArrayList<String>()).nargs("+")
+				.help("Only include hosts that match all of these labels. Labels need to be in the format "
+						+ "key=value.");
+	}
+
+	@Override
+	int run(final Namespace options, final HeliosClient client, final PrintStream out, final boolean json,
+			final BufferedReader stdin) throws ExecutionException, InterruptedException {
+		final String pattern = options.getString(patternArg.getDest());
+		final List<String> hosts = FluentIterable.from(client.listHosts().get()).filter(containsPattern(pattern))
```

Comment:
@feroozkhanchintu Thanks for your PR.

Always calling `HeliosClient.listHosts()` might slow down this command if there are a lot of hosts but only a few hosts that match a specified pattern or selector. `HeliosClient.listHosts(pattern, selectors)` makes the server filter and results in less data being transmitted over the wire. (The ZK operation is still to get all the child nodes of `/config/hosts` though.)

I'm also a fan of keeping as much logic on the server side as possible.